fix issue1: update for current core_kernel and lwt

dbrumley opened the pull request:
Fixes labeled arguments for `bind` compilation error in issue 1 for modern core_kernel and lwt.

Commit:
fix issue1: update for current core_kernel and lwt

dbrumley commented:
that's weird. the travis build problem is the one I thought the patch fixed...at least on my system.

ivg commented on lib/core_lwt.ml:
```diff
@@ -9,7 +9,7 @@ module Std = struct
     module Main = Lwt_main
     module Unix = Lwt_unix
-    module Chan = Lwt_chan
+    module Chan = Lwt_io
```
This substitution is not valid, `Chan` is a module alias for `Lwt_chan`, and the same is true for other modules. We just moving here from the pseudo `Lwt_` namespace, to a modular `Lwt.`, e.g., `Lwt_xxx` is now `Lwt.Xxx`.

dbrumley replied:
During build without the change:
```
File "lib/core_lwt.ml", line 12, characters 18-26:
Warning 3: deprecated: module Lwt_chan
 Use module Lwt_io.
```
Do you want to rename `Chan` to `Io` for the substitution?

ivg replied:
We can either drop Chan at all, or keep it deprecated. I think that we can easily just remove it.

ivg reviewed (changes requested):
You need to update the opam file and constrain the lower bound of the core_kernel version. It should be something above v0.9. 

The build fails because it grabs core.113.00 that doesn't use a labeled parameter for the `bind` function.

Commit:
bumped core_kernel version requirement

ivg commented on opam:
```diff
@@ -25,7 +25,7 @@ depends: [
     "base-unix"
     "base-threads"
-    "core_kernel"
+    "core_kernel" { >= "0.9.0"}
```
it should be `v0.9`, notice the leading `v`.

dbrumley replied:
i was wondering that and was going to let travis tell me, but this is quicker. (It did build on my system).

dbrumley replied:
I'll change that after the above is answered re: `Chan`.

Commit:
Fix version req. for core_kernel, removed deprecated interface.

Bumped version to require core_kernel v0.9.0.
Removed `Lwt_chan`, which is deprecated for `Lwt_io`.

ivg commented:
OK, now it fails because the OCAML compiler version is too old. Although it is set to the "latest", apparently the latest value is hardcoded somewhere in the scripts upstream, so it is no longer the latest. My suggestion is to set it explicitly to 4.05.0 in travis.yml (see the environment variable OCAML_VERSION in the env stanza).

Commit:
Set explicitly ocaml to 4.03.0 instead of latest.

Although travis is set to "latest", it seems to be pulling in an older
version. Set to 4.03.0, which is what is working on my system.

dbrumley commented:
Set to 4.03.0 as that's what is working on my system. Fingers crossed.

dbrumley commented:
Alright, I'm confused here. Travis is failing with:
```
  - core_kernel>=v0.9.0 is not available because your system doesn't comply with ocaml-version >= "4.04.1".
```

Yet on my system:
```
(py36) Davids-MBP:core-lwt dbrumley$ opam info core_kernel  | head -2
             package: core_kernel
             version: v0.9.0
(py36) Davids-MBP:core-lwt dbrumley$ ocamlc --version
4.03.0
```

I could bump to 4.05.0 as originally suggested, but would love to know why it seems to work fine with 4.03.0 on my system with v0.9.0 core_kernel.

ivg commented:
I think that it is just a failure in OPAM, because aspcud is not installed and in its absence a heuristic is used ... that fails more often than succeeds. The ASP problem, that OPAM is trying to solve, is undecidable in general, and given that specifications in the opam-repository is an approximation, that is both over-constrained (for some packages) and under-constrained (for other packages) the solution is often far from perfect. 

The solution that we see here is really funny, I've greped the opam-repository for any occurrences of  4.04.1 and the only package that states such constraint is some `horned_worm` package, that is not used by any other packages at all. So where this constraint did come from, I have no idea. Probably, it is just some widening.

So the main problem is that the `aspcud` package is not installed, and the proper solution would be to add it into the `packages` section in the `.travis.yml`

Commit:
added aspcud to solve travis build problem to packages

ivg commented:
damn, now I'm frustrated)) Of course, I can't reproduce it on my system. Though I have exactly the same setup as on Travis.

ivg committed:
enables OPAM_SWITCH

it looks like that we still install the system compiler, as the travis log contains: 
```
opam init -a git://github.com/ocaml/opam-repository --comp=system
```

I hope that this option will force it to install the compiler, instead of using the system one.

dbrumley commented:
Is it possible opam package information is cached? Couldn't tell from build if it was a new switch or installed and updated from scratch. ??????

ivg commented:
> Is it possible opam package information is cached?

No, it's not, caching is disabled for this repository. This was the first thing I've checked. 

So my last commit fixed this issue, though we have the new one:
```
Error: The implementation lib/core_lwt_pool.ml
# [...]
3m#            ?validate:('a -> bool Lwt.t) ->
#            ?check:('a -> (bool -> unit) -> unit) ->
#            ?dispose:('a -> unit Lwt.t) -> (unit -> 'a Lwt.t) -> 'a t
#        is not included in
#          val create :
#            int ->
#            ?check:('a -> (bool -> unit) -> unit) ->
#            ?validate:('a -> bool Lwt.t) -> (unit -> 'a Lwt.t) -> 'a t
```

Apparently, it is because you have a different version of Lwt on your system, and in version that is used in Travis there is a new parameter called `dispose` and no longer have an unnamed parameter of type `int` (whatever it is). 

The solution would be to add explicit constraint to the lwt version. You can use the `=` constraint it put the version that you're using, alternatively, you can use a range, e.g., `{ >= "2.0.0" & < "3.0.0"}`. Anton respects semantic versioning, so if your code compiles with the 2.0 then it should compile with  all 2.x. (I'm not sure are you using 1.x, or 2.x, or 3.x so make sure that choose the proper version)

Commit:
Updated to lwt-3.0.0 API and updated opam dependencies

Commit:
Merge branch 'master' of github.com:dbrumley/core-lwt

Commit:
Updated core_kernel requirement to work with bap dev and 1.3

Commit:
Changed dependencies to be more lax on lwt

Commit:
bumped core-kernel dep to 113.33.01

Commit:
fixed opam to work with bap development

dbrumley commented:
Alright, seems the original issue is with bap using v0.9.0 of core_kernel, which is used by the current development version of bap.  

The current version of core-lwt works fine with bap-1.3 as in opam, as that depends on core_kernel 1113.33.xx.  (There was an error in my build where it was using an outdated `setup.data` with a bad dependency.)

I'm going to test a build just to verify everything works with bap 1.3 and core_kernel v0.9.0. It's unclear whether this patch should be pushed now or wait even if it builds.

ivg reviewed (approved):
👍
